### PR TITLE
feat: support alter index

### PR DESCRIPTION
**Pull request by yah01**
related: https://github.com/milvus-io/milvus/issues/21866

### CODE REVIEW

File: pymilvus/client/grpc_handler.py [hunk 1]
```diff
@@ -921,6 +921,24 @@ def _check():
...
+        **kwargs,
+    ):
+        check_pass_param(collection_name=collection_name)
+        request = Prepare.alter_index_request(collection_name, index_name, extra_params)
+
+        rf = self._stub.AlterIndex.future(request, timeout=timeout)
+        response = rf.result()
+        status = response.status
+        check_status(status)
+        return Status(status.code, status.reason)
```
Comment (XuanYang-cn) [inline pymilvus/client/grpc_handler.py:941]: don't need to return Status like create_index, just check is enough

File: pymilvus/client/stub.py [hunk 2]
```diff
@@ -652,6 +652,29 @@ def create_index(self, collection_name, field_name, params, timeout=None, **kwar
...
+        :param timeout: An optional duration of time in seconds to allow for the RPC. When timeout
+                        is set to None, client waits until server responses or errors occur.
+        :type  timeout: float
+
+        :return: None
+        :rtype: NoneType
+        """
+        with self._connection() as handler:
+            return handler.alter_index(collection_name, index_name, extra_params, timeout=timeout)
+
```
Comment (XuanYang-cn) [inline pymilvus/client/stub.py:677]: stub.py is deprecated, no need to impl here, please remove the code

File: pymilvus/orm/collection.py [hunk 3]
```diff
@@ -1349,6 +1349,55 @@ def create_index(
...
+            >>> collection = Collection("test_collection_alter_index", schema)
+            >>> index_params = {
+            ...     "index_type": IndexType.IVF_FLAT,
+            ...     "metric_type": MetricType.L2,
+            ...     "params": {"nlist": 128}
+            ... }
+            >>> collection.create_index("films", index_params, index_name="idx")
+            Status(code=0, message='')
+            >>> collection.alter_index("idx", {"mmap.enabled": True)
+            Status(code=0, message='')
```
Comment (XuanYang-cn) [inline pymilvus/orm/collection.py:1396]: Please remove the return, no expections means success

**Review by XuanYang-cn — CHANGES_REQUESTED** [review]

File: pymilvus/client/prepare.py [hunk 4]
```diff
@@ -726,6 +726,20 @@ def dump(tv: Dict):
 
         return index_params
 
+    @classmethod
+    def alter_index_request(cls, collection_name: str, index_name: str, extra_params: dict):
+        def dump(tv: Dict):
+            if isinstance(tv, dict):
+                return ujson.dumps(tv)
+            return str(tv)
```
Comment (XuanYang-cn) [inline pymilvus/client/prepare.py:734]: Once line is more pythonic and has more readibility
```python
value = ujson.dumps(extra_params) if isinstance(extra_params, dict) else str(extra_params)
```

File: pymilvus/client/grpc_handler.py [hunk 5]
```diff
@@ -922,6 +922,23 @@ def _check():
...
+    def alter_index(
+        self,
+        collection_name: str,
+        index_name: str,
+        extra_params: dict,
+        timeout: Optional[float] = None,
+        **kwargs,
+    ):
+        check_pass_param(collection_name=collection_name)
+        request = Prepare.alter_index_request(collection_name, index_name, extra_params)
```
Comment (XuanYang-cn) [inline pymilvus/client/grpc_handler.py:935]: please check index_name a valid string and extra_params not None

File: pymilvus/orm/collection.py [hunk 6]
```diff
@@ -1349,6 +1349,48 @@ def create_index(
...
+            ... ])
+            >>> collection = Collection("test_collection_alter_index", schema)
+            >>> index_params = {
+            ...     "index_type": IndexType.IVF_FLAT,
+            ...     "metric_type": MetricType.L2,
+            ...     "params": {"nlist": 128}
+            ... }
+            >>> collection.create_index("films", index_params, index_name="idx")
+            Status(code=0, message='')
+            >>> collection.alter_index("idx", {"mmap.enabled": True)
```
Comment (XuanYang-cn) [inline pymilvus/orm/collection.py:1389]: missing right }

File: pymilvus/client/grpc_handler.py (hunk 5 above)
Comment (yah01) [reply pymilvus/client/grpc_handler.py:935]: any criterion to check `index_name`?

**Pull request merged**